### PR TITLE
feat(harmonytypes): introduce Photo, Minithumbnail, and ContentPhotos

### PR DESCRIPTION
**Pull request by pontaoski**

### CODE REVIEW

File: harmonytypes/v1/types.proto [hunk 1]
```diff
@@ -115,6 +115,45 @@ message Attachment {
...
+  // The size of the photo.
+  int32 file_size = 3;
+  // The height of the photo, in pixels.
+  int32 height = 6;
+  // The width of the photo, in pixels.
+  int32 width = 7;
+  // The photo's caption.
+  string caption = 4;
+  // A thumbnail representing the photo.
+  Minithumbnail minithumbnail = 5;
```
Comment (yusdacra) [inline harmonytypes/v1/types.proto:154]:
```suggestion
  // The photo's caption.
  string caption = 8;
  // A thumbnail representing the photo.
  Minithumbnail minithumbnail = 9;
```

File: harmonytypes/v1/types.proto [hunk 2]
```diff
@@ -115,6 +115,45 @@ message Attachment {
   string caption = 5;
 }
 
+/*
+  Minithumbnail is an extremely low-quality JPEG thumbnail.
+
+  The resolution is usually no larger than 64x64.
```
Comment (yusdacra) [inline harmonytypes/v1/types.proto:121]: 64x64 is too small, and images can vary resolution ratios. We should put a file size limit instead like 100KB.

**Review by yusdacra — CHANGES_REQUESTED** [review]

File: harmonytypes/v1/types.proto (hunk 2 above)
Comment (pontaoski) [reply harmonytypes/v1/types.proto:121]: 100KB is far too big; considering that the JPEG data is inlined in the message itself.

The purpose of a minithumbnail is supposed to be something that the client can use to get a blob of colours to display blurrily, like so:

Comment (pontaoski) [reply harmonytypes/v1/types.proto:121]: Screenshot_20210813-210123

Comment (yusdacra) [reply harmonytypes/v1/types.proto:121]: Well it should at least match the width and height of the image, so we can remove the width and height from the Minithumbnail and move the data field into Photo itself.

Comment (pontaoski) [reply harmonytypes/v1/types.proto:121]: no, the minithumbnail needs to be extremely small, otherwise you bloat the message with too much photo data

Comment (yusdacra) [reply harmonytypes/v1/types.proto:121]: Well, if we want to display it in message history, once the actual photo loads, the messages will jump up and down. So the width and height should match the actual images as in it's ratio, ie. if it's 16:9 it should stay 16:9.

Comment (pontaoski) [reply harmonytypes/v1/types.proto:121]: 64x64 is the recommended maximum; doesn't mean messages have to be 1:1 ratio

**Review by yusdacra — APPROVED** [review]

File: harmonytypes/v1/types.proto [hunk 3]
```diff
@@ -123,12 +169,14 @@ message Metadata {
 message ContentText { string content = 1; }
 message ContentEmbed { Embed embeds = 1; }
 message ContentFiles { repeated Attachment attachments = 1; }
+message ContentPhotos { repeated Attachment photos = 1; }
```
Comment (yusdacra) [inline harmonytypes/v1/types.proto:172]:
```suggestion
message ContentPhotos { repeated Photo photos = 1; }
```

**Review by yusdacra — CHANGES_REQUESTED** [review]

**Pull request merged**